Code/Engine/RendererCore/Meshes/Implementation/CustomMeshComponent.cpp

```diff
@@ -0,0 +1,379 @@
+  }
+
+  return EZ_FAILURE;
+}
+
+ezDynamicMeshBufferResourceHandle ezCustomMeshComponent::CreateMeshResource(ezUInt32 uiNumVertices, ezUInt32 uiNumPrimitives, ezUInt32 uiNumIndices, bool b32BitIndices /*= true*/)
+{
+  ezDynamicMeshBufferResourceDescriptor desc;
+  desc.m_Topology = ezGALPrimitiveTopology::Triangles;
+  desc.m_uiNumVertices = uiNumVertices;
+  desc.m_uiNumPrimitives = uiNumPrimitives;
+  desc.m_uiNumIndices16 = b32BitIndices ? 0 : uiNumIndices;
+  desc.m_uiNumIndices32 = b32BitIndices ? uiNumIndices : 0;
+
+  ezStringBuilder sGuid;
+  sGuid.Format("CustomMesh_{}", s_iCustomMeshResources.Increment());
+
+  m_hDynamicMesh = ezResourceManager::CreateResource<ezDynamicMeshBufferResource>(sGuid, std::move(desc));
+
+  InvalidateCachedRenderData();
+
+  return m_hDynamicMesh;
+}
+
+void ezCustomMeshComponent::SetMeshResource(const ezDynamicMeshBufferResourceHandle& hMesh)
+{
+  m_hDynamicMesh = hMesh;
+  InvalidateCachedRenderData();
+}
+
+void ezCustomMeshComponent::SetBounds(const ezBoundingBoxSphere& bounds)
+{
+  m_Bounds = bounds;
+  TriggerLocalBoundsUpdate();
+}
+
+void ezCustomMeshComponent::SetMaterial(const ezMaterialResourceHandle& hMaterial)
+{
+  m_hMaterial = hMaterial;
+  InvalidateCachedRenderData();
+}
+
+ezMaterialResourceHandle ezCustomMeshComponent::GetMaterial() const
+{
+  return m_hMaterial;
+}
+
+void ezCustomMeshComponent::SetMaterialFile(const char* szMaterial)
+{
+  ezMaterialResourceHandle hResource;
+
+  if (!ezStringUtils::IsNullOrEmpty(szMaterial))
+  {
+    hResource = ezResourceManager::LoadResource<ezMaterialResource>(szMaterial);
+  }
+
+  m_hMaterial = hResource;
+}
+
+const char* ezCustomMeshComponent::GetMaterialFile() const
+{
+  if (!m_hMaterial.IsValid())
+    return "";
+
+  return m_hMaterial.GetResourceID();
+}
+
+void ezCustomMeshComponent::SetColor(const ezColor& color)
+{
+  m_Color = color;
+
+  InvalidateCachedRenderData();
+}
+
+const ezColor& ezCustomMeshComponent::GetColor() const
+{
+  return m_Color;
+}
+
+void ezCustomMeshComponent::OnMsgSetMeshMaterial(ezMsgSetMeshMaterial& msg)
+{
+  SetMaterial(msg.m_hMaterial);
+}
+
+void ezCustomMeshComponent::OnMsgSetColor(ezMsgSetColor& msg)
+{
+  msg.ModifyColor(m_Color);
+
+  InvalidateCachedRenderData();
+}
+
+void ezCustomMeshComponent::SetUsePrimitiveRange(ezUInt32 uiFirstPrimitive /*= 0*/, ezUInt32 uiNumPrimitives /*= ezMath::MaxValue<ezUInt32>()*/)
+{
+  m_uiFirstPrimitive = uiFirstPrimitive;
+  m_uiNumPrimitives = uiNumPrimitives;
+}
+
+void ezCustomMeshComponent::OnMsgExtractRenderData(ezMsgExtractRenderData& msg) const
+{
+  if (!m_hDynamicMesh.IsValid() || !m_hMaterial.IsValid())
+    return;
+
+  ezResourceLock<ezDynamicMeshBufferResource> pMesh(m_hDynamicMesh, ezResourceAcquireMode::BlockTillLoaded);
+
+  ezCustomMeshRenderData* pRenderData = ezCreateRenderDataForThisFrame<ezCustomMeshRenderData>(GetOwner());
+  {
+    pRenderData->m_GlobalTransform = GetOwner()->GetGlobalTransform() * pRenderData->m_GlobalTransform;
```

Comment:
pRenderData->m_GlobalTransform is identity at this point so it doesn't make sense to multiply with it.